cpp/src/arrow/array/builder_base.h

```diff
@@ -145,6 +168,11 @@ class ARROW_EXPORT ArrayBuilder {
   virtual Status AppendScalar(const Scalar& scalar, int64_t n_repeats);
   virtual Status AppendScalars(const ScalarVector& scalars);
 
+  /// \brief Append a scalar value from an optional or null if it has no value.
```

Comment:
Oh right. The comment on this one shouldn't be like the others.

> Also, TBH, I'm not sure I understand the point of this API. Are there situations where this would be useful?

I'm planning to use it in the REE array builder. Representing the current open run value with a `shared_ptr<Scalar>` and length.

Reply:
Going through a `shared_ptr<Scalar>` would be too costly IMHO.

Reply:
Ok, I will keep this in the builder code then.